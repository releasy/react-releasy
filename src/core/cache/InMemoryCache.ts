import { Variables, GraphQLResponse } from 'relay-runtime';
import RelayQueryResponseCache from 'relay-runtime/lib/RelayQueryResponseCache.js';

import { CacheInterface } from '../Config';

type InMemoryCacheProps = {
  ttl?: number,
  size?: number,
};

class InMemoryCache implements CacheInterface {
  store: RelayQueryResponseCache;

  constructor(props?: InMemoryCacheProps) {
    this.store = new RelayQueryResponseCache({
      size: props && props.size ? props.size : 250,
      ttl: props && props.ttl ? props.ttl : 60 * 1000 * 5, // default = 5 minutes
    });
  }

  get = (queryID: string, variables: Variables): GraphQLResponse => {
    return this.store.get(queryID, variables);
  }

  set = (queryID: string, variables: Variables, payload: GraphQLResponse): void => {
    this.store.set(queryID, variables, payload);
  }

  clear = (): void => {
    this.store.clear();
  }
}

export default InMemoryCache;
