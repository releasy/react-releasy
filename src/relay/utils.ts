import { CacheConfig, RequestNode } from 'relay-runtime';

export const isMutation = (request: RequestNode) => request.operationKind === 'mutation';

export const mustForceFetch = (cacheConfig: CacheConfig) => !!(cacheConfig && cacheConfig.force);
