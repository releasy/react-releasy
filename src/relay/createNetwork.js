// @flow
import { Network } from 'relay-runtime';
import RelayNetworkLogger from 'relay-runtime/lib/RelayNetworkLogger';

import type { CacheConfig, RequestNode, UploadableMap, Variables } from 'relay-runtime';

import { isMutation, mustForceFetch } from './utils';

import type { Config } from '../index';

const createNetwork = (config: Config): Network => {
  const fetchFn = config.networkLogger
    ? RelayNetworkLogger.wrapFetch(config.link.fetch)
    : config.link.fetch;

  return Network.create(
    async (
      request: RequestNode,
      variables: Variables,
      cacheConfig: CacheConfig,
      uploadables?: UploadableMap,
    ) => {
      const queryID = request.text;

      // no cache configured, just fetch the data from server
      if (!config.cache) {
        return fetchFn(request, variables, cacheConfig, uploadables);
      }

      // on mutations, clear cache and then fetch the data
      if (isMutation(request)) {
        config.cache.clear();
        return fetchFn(request, variables, cacheConfig, uploadables);
      }

      // verify if data exists on cache
      const fromCache = config.cache.get(queryID, variables);
      if (fromCache !== null && !mustForceFetch(cacheConfig)) {
        return fromCache;
      }

      const fromServer = await fetchFn(request, variables, cacheConfig, uploadables);
      if (fromServer) {
        config.cache.set(queryID, variables, fromServer);
      }

      return fromServer;
    }
  );
};

export default createNetwork;
