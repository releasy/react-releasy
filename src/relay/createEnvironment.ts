import { installRelayDevTools } from 'relay-devtools';
import { Environment, RecordSource, Store } from 'relay-runtime';

import Config from '../core/Config';
import createNetwork from './createNetwork';

const createEnvironment = (config: Config): Environment => {
  if (config.devTools) {
    installRelayDevTools();
  }

  const network = createNetwork(config);

  const source = new RecordSource();
  const store = new Store(source);

  return new Environment({
    network,
    store,
  });
};

export default createEnvironment;
