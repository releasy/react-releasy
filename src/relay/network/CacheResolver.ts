import {
  CacheConfig,
  RequestNode,
  UploadableMap,
  Variables,
  GraphQLResponse,
  Sink,
} from 'relay-runtime';

import Config from '../../core/Config';
import { isMutation, mustForceFetch } from '../utils';
import { NetworkResolverInterface } from './createNetwork';

class CacheResolver implements NetworkResolverInterface {
  config: Config;

  constructor(config: Config) {
    this.config = config;
  }

  resolve(
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables: UploadableMap,
    sink: Sink<GraphQLResponse>,
  ) {
    if (isMutation(request)) {
      sink.next();
      return;
    }

    if (!this.config.cache || mustForceFetch(cacheConfig)) {
      sink.next();
      return;
    }

    const data = this.config.cache.get(request.text, variables);
    if (!data) {
      sink.next();
      return;
    }

    sink.next(data);
    sink.complete();
  }
}

export default CacheResolver;
