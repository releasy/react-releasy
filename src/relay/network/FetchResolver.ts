import {
  CacheConfig,
  RequestNode,
  UploadableMap,
  Variables,
  GraphQLResponse,
  Sink,
} from 'relay-runtime';
import RelayNetworkLogger from 'relay-runtime/lib/RelayNetworkLogger';

import Config from '../../core/Config';
import { isMutation } from '../utils';
import { NetworkResolverInterface } from './createNetwork';

class FetchResolver implements NetworkResolverInterface {
  config: Config;

  constructor(config: Config) {
    this.config = config;
  }

  async resolve(
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables: UploadableMap,
    sink: Sink<GraphQLResponse>,
  ) {
    const fetchFn = this.config.networkLogger
      ? RelayNetworkLogger.wrapFetch(this.config.link.fetch)
      : this.config.link.fetch;

    try {
      if (isMutation(request) && !!this.config.cache) {
        this.config.cache.clear();
      }

      const data = await fetchFn(request, variables, cacheConfig, uploadables);

      if (!isMutation(request) && !!this.config.cache) {
        this.config.cache.set(request.text, variables, data);
      }

      sink.next(data);
      sink.complete();
    } catch (error) {
      sink.error(error);
      sink.complete();
    }
  }
}

export default FetchResolver;
