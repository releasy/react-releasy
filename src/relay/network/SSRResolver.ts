import {
  CacheConfig,
  RequestNode,
  UploadableMap,
  Variables,
  GraphQLResponse,
  Sink,
} from 'relay-runtime';

import { IS_NODE, WINDOW_DATA_KEY } from '../../utils/constants';
import InMemoryCache from '../../core/cache/InMemoryCache/InMemoryCache';
import { isMutation } from '../utils';
import { NetworkResolverInterface } from './createNetwork';

class SSRResolver implements NetworkResolverInterface {
  cache: InMemoryCache;

  constructor() {
    this.cache = new InMemoryCache();

    const ssrData = (!IS_NODE && window[WINDOW_DATA_KEY]) || [];
    ssrData.forEach(({ operation, data }) => {
      this.cache.set(operation.node.text, operation.variables, { data });
    });
  }

  resolve = (
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables: UploadableMap,
    sink: Sink<GraphQLResponse>,
  ) => {
    if (isMutation(request)) {
      sink.next();
      return;
    }

    const data = this.cache.get(request.text, variables);
    if (!data) {
      sink.next();
      return;
    }

    this.cache.set(request.text, variables, null);

    sink.next(data);
    sink.complete();
  }
}

export default SSRResolver;
