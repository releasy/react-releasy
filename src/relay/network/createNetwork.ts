import {
  Network,
  CacheConfig,
  RequestNode,
  UploadableMap,
  Variables,
  Observable,
  Sink,
  GraphQLResponse,
} from 'relay-runtime';

import Config from '../../core/Config';
import CacheResolver from './CacheResolver';
import FetchResolver from './FetchResolver';
import SSRResolver from './SSRResolver';

export interface NetworkResolverInterface {
  resolve(
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables: UploadableMap,
    sink: Sink<GraphQLResponse>,
  );
}

const createNetwork = (config: Config): Network => {
  const resolvers: NetworkResolverInterface[] = [
    new SSRResolver(),
    new CacheResolver(config),
    new FetchResolver(config),
  ];

  return Network.create(
    (
      request: RequestNode,
      variables: Variables,
      cacheConfig: CacheConfig,
      uploadables?: UploadableMap,
    ) => new Promise((resolve, reject) => {
      let sink: Sink<GraphQLResponse>;
      let data: GraphQLResponse;
      let currentResolver = -1;

      Observable.create((_sink) => {
        sink = _sink;
        _sink.next();
      }).subscribe({
        next: (_data: GraphQLResponse) => {
          if (data) {
            return;
          }

          if (_data) {
            data = _data;
          }

          currentResolver += 1;

          const resolver = resolvers[currentResolver];
          if (!resolver) {
            sink.complete();
            return;
          }

          resolver.resolve(request, variables, cacheConfig, uploadables, sink);
        },
        complete: () => resolve(data),
        error: reject,
      });
    }),
  );
};

export default createNetwork;
