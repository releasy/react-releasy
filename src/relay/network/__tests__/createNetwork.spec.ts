import { createConfig } from '../../../../tests/fixtures/core';
import {
  createCacheConfig,
  createMutationRequest,
  createMutationResponse,
  createQueryRequest,
  createQueryResponse,
} from '../../../../tests/fixtures/relay';
import { delay } from '../../../../tests/utils';

import { WINDOW_DATA_KEY } from '../../../utils/constants';
import { getHash } from '../../../relay/utils';
import InMemoryCache from '../../../core/cache/InMemoryCache/InMemoryCache';
import createNetwork from '../createNetwork';
import CacheResolver from '../CacheResolver';
import FetchResolver from '../FetchResolver';
import SSRResolver from '../SSRResolver';

beforeEach(() => {
  jest.clearAllMocks();
});

it('should create default Network', () => {
  const config = createConfig();

  const network = createNetwork(config);

  expect(network).toMatchSnapshot();
});

it('should create Network with networkLogger', () => {
  const config = createConfig({ networkLogger: true });

  const network = createNetwork(config);

  expect(network).toMatchSnapshot();
});

it('should create Network with cache', () => {
  const config = createConfig({ cache: new InMemoryCache() });

  const network = createNetwork(config);

  expect(network).toMatchSnapshot();
});

it('should resolve with SSR', async () => {
  const ssrResolver = jest.spyOn(SSRResolver.prototype, 'resolve');
  const cacheResolver = jest.spyOn(CacheResolver.prototype, 'resolve');
  const fetchResolver = jest.spyOn(FetchResolver.prototype, 'resolve');

  const request = createQueryRequest();
  const variables = {};
  const cacheConfig = createCacheConfig();
  const response = createQueryResponse();

  window[WINDOW_DATA_KEY] = { [getHash(request, variables)]: response.data };

  // make sure this is being created after releasy data is populated
  const config = createConfig();
  const network = createNetwork(config);

  (network as any).execute(request, variables, cacheConfig);

  expect(ssrResolver).toHaveBeenCalled();
  expect(cacheResolver).not.toHaveBeenCalled();
  expect(fetchResolver).not.toHaveBeenCalled();
});

it('should resolve with cache', async () => {
  const ssrResolver = jest.spyOn(SSRResolver.prototype, 'resolve');
  const cacheResolver = jest.spyOn(CacheResolver.prototype, 'resolve');
  const fetchResolver = jest.spyOn(FetchResolver.prototype, 'resolve');

  const config = createConfig({ cache: new InMemoryCache() });
  const network = createNetwork(config);

  const request = createQueryRequest();
  const variables = {};
  const cacheConfig = createCacheConfig();

  global.fetch.mockResponseOnce(
    JSON.stringify(createQueryResponse()),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  (network as any).execute(request, variables, cacheConfig);

  await delay(100);

  (network as any).execute(request, variables, cacheConfig);

  // first call will bypass ssr and cache
  expect(ssrResolver).toHaveBeenCalledTimes(2);
  expect(cacheResolver).toHaveBeenCalledTimes(2);
  expect(fetchResolver).toHaveBeenCalledTimes(1);
});

it('shouldn\'t resolve with cache and also clear the cache on mutations', async () => {
  const ssrResolver = jest.spyOn(SSRResolver.prototype, 'resolve');
  const cacheResolver = jest.spyOn(CacheResolver.prototype, 'resolve');
  const fetchResolver = jest.spyOn(FetchResolver.prototype, 'resolve');

  const config = createConfig({ cache: new InMemoryCache() });
  const network = createNetwork(config);

  const queryRequest = createQueryRequest();
  const queryVariables = {};
  const queryCacheConfig = createCacheConfig();

  global.fetch.mockResponseOnce(
    JSON.stringify(createQueryResponse()),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  (network as any).execute(queryRequest, queryVariables, queryCacheConfig);

  await delay(100);

  expect(config.cache.size()).toBe(1);

  const mutationRequest = createMutationRequest();
  const mutationVariables = {};
  const mutationCacheConfig = createCacheConfig();

  global.fetch.mockResponseOnce(
    JSON.stringify(createMutationResponse()),
    {
      status: 200,
      headers: {
        'content-type': '*/*',
      },
    },
  );

  (network as any).execute(mutationRequest, mutationVariables, mutationCacheConfig);

  expect(ssrResolver).toHaveBeenCalledTimes(2);
  expect(cacheResolver).toHaveBeenCalledTimes(2);
  expect(fetchResolver).toHaveBeenCalledTimes(2);

  expect(config.cache.size()).toBe(0);
});
