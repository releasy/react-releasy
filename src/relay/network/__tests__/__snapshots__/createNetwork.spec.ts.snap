// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create Network with cache 1`] = `
Object {
  "execute": [Function],
  "executeWithEvents": [Function],
}
`;

exports[`should create Network with networkLogger 1`] = `
Object {
  "execute": [Function],
  "executeWithEvents": [Function],
}
`;

exports[`should create default Network 1`] = `
Object {
  "execute": [Function],
  "executeWithEvents": [Function],
}
`;
